lib/src/auth.dart

```diff
@@ -264,3 +282,102 @@ class ReadOnlyAuthenticator extends Authenticator {
+  /// The value of [state] will be returned via the redirect Uri and should be
+  /// verified against the original value of [state] to ensure the app access
+  /// response corresponds to the correct request. [duration] indicates whether
+  /// or not a permanent token is needed for the client, and can take the value
+  /// of either 'permanent' (default) or 'temporary'. If [compactLogin] is true,
+  /// then the Uri will link to a mobile-friendly Reddit authentication screen.
+  Uri url(List<String> scopes, String state,
+      {String duration = 'permanent', bool compactLogin = false}) {
+    // TODO(bkonyi) do we want to add the [implicit] flag to the argument list?
+    if (scopes == null) {
+      // scopes cannot be null.
+      throw new DRAWAuthenticationError('Parameter scopes cannot be null.');
+    }
+    Uri redditAuthUri =
+        _grant.getAuthorizationUrl(_redirect, scopes: scopes, state: state);
+    if (redditAuthUri == null) {
+      // TODO(bkonyi) throw meaningful exception.
+      assert(false);
+    }
+    // getAuthorizationUrl returns a Uri which is missing the duration field, so
+    // we need to add it here.
+    Map queryParameters = new Map.from(redditAuthUri.queryParameters);
+    queryParameters[kDurationKey] = duration;
+    redditAuthUri = redditAuthUri.replace(queryParameters: queryParameters);
+    if (compactLogin) {
+      String path = redditAuthUri.path;
+      assert(path.endsWith('?'), 'The path should end with "authorize?"');
+      path = path.substring(0, path.length - 1) + r'.compact?';
+      redditAuthUri = redditAuthUri.replace(path: path);
+    }
+    return redditAuthUri;
+  }
+
+  /// Authorizes the current [Authenticator] instance using the code returned
+  /// from Reddit after the user has authenticated.
+  ///
+  /// [code] is the value passed as a query parameter to `redirect`. This value
+  /// must be parsed from the request made to `redirect` before being passed to
+  /// this method.
+  Future authorize(String code) async {
+    if (code == null) {
+      // code cannot be null.
+      throw new DRAWAuthenticationError('Parameter code cannot be null.');
+    }
+    _client = await _grant.handleAuthorizationCode(code);
+  }
+
+  /// Initiates the authorization flow. This method should populate a
+  /// [Map<String,String>] with information needed for authentication, and then
+  /// call [_requestToken] to authenticate.
+  @override
+  Future _authenticationFlow() async {
+    throw new UnimplementedError(
+        '_authenticationFlow is not used in WebAuthenticator.');
+  }
+
+  /// Request a new access token from the Reddit API. Throws a
+  /// [DRAWAuthenticationError] if the [Authenticator] is not yet initialized.
+  @override
+  Future refresh() async {
+    if (_client == null) {
+      throw new DRAWAuthenticationError(
+          'cannot refresh uninitialized Authenticator.');
+    }
+    _client = await _client.refreshCredentials();
+  }
+
+  Uri _redirect;
```

Comment:
nit: Fields are generally declared at the top of your class before constructors.